File: z_lc/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy = ListNode(0, head);
        ListNode *back = &dummy, *front = &dummy;
        for(int _=0; _<n+1; ++_) {
            front = front->next;
        }
        while(front) {
            back = back->next;
            front = front->next;
        }
        back->next = back->next->next;
        return dummy.next;
    }
};

File: z_lc/19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution: 
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(next = head)
        back, front = dummy, dummy
        for i in range(n+1):
            front = front.next
        while front:
            back, front = back.next, front.next
        back.next = back.next.next
        return dummy.next